lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobDriver.java

```diff
@@ -230,41 +234,40 @@ private CLRBufferedLogHandler getCLRBufferedLogHandler() {
   }
 
   private void submitEvaluator(final AllocatedEvaluator eval, final EvaluatorProcess process) {
-    synchronized (JobDriver.this) {
-      eval.setProcess(process);
-      LOG.log(Level.INFO, "Allocated Evaluator: {0}, total running running {1}",
-          new Object[]{eval.getId(), JobDriver.this.contexts.size()});
-      if (JobDriver.this.handlerManager.getAllocatedEvaluatorHandler() == 0) {
-        throw new RuntimeException("Allocated Evaluator Handler not initialized by CLR.");
-      }
-      final AllocatedEvaluatorBridge allocatedEvaluatorBridge =
-          this.allocatedEvaluatorBridgeFactory.getAllocatedEvaluatorBridge(eval, this.nameServerInfo);
-      allocatedEvaluatorBridges.put(allocatedEvaluatorBridge.getId(), allocatedEvaluatorBridge);
-      NativeInterop.clrSystemAllocatedEvaluatorHandlerOnNext(
-          JobDriver.this.handlerManager.getAllocatedEvaluatorHandler(), allocatedEvaluatorBridge, this.interopLogger);
+    eval.setProcess(process);
+    LOG.log(Level.FINE, "Allocated Evaluator: {0}, total running running {1}, time: {2}",
+        new Object[]{eval.getId(), JobDriver.this.contexts.size(), dateFormat.format(new Date())});
+    final long handler = JobDriver.this.handlerManager.getAllocatedEvaluatorHandler();
+    if (handler == 0) {
+      throw new RuntimeException("Allocated Evaluator Handler not initialized by CLR.");
     }
+    final AllocatedEvaluatorBridge allocatedEvaluatorBridge =
+        this.allocatedEvaluatorBridgeFactory.getAllocatedEvaluatorBridge(eval, this.nameServerInfo);
+    allocatedEvaluatorBridges.put(allocatedEvaluatorBridge.getId(), allocatedEvaluatorBridge);
```

Comment:
Maybe it makes sense to check the return value here to make sure the map did not contain this entry before.

Reply:
Changed to putIfAbsent to make it atomic.